front/components/agent_builder/capabilities/knowledge/KnowledgeConfigurationSheet.tsx

```diff
@@ -26,147 +28,198 @@ import {
+      </MultiPageSheetTrigger>
+      <KnowledgeConfigurationSheetContent
+        config={config}
+        action={action}
+        onSave={handleSave}
+        control={control}
+        handleSubmit={handleSubmit}
+        isOpen={open}
+      />
     </MultiPageSheet>
   );
 }
 
 interface KnowledgeConfigurationSheetContentProps {
-  config: CapabilityConfig;
+  config: CapabilityConfig | null;
   action?: AgentBuilderAction;
   onSave: (
     formData: CapabilityFormData,
-    dataSourceConfigurations: DataSourceViewSelectionConfigurations
+    dataSourceConfigurations: DataSourceViewSelectionConfigurations,
+    configToUse: CapabilityConfig
   ) => void;
-  onClose: () => void;
   control: Control<CapabilityFormData>;
   handleSubmit: (
     fn: (data: CapabilityFormData) => void
   ) => (e?: React.BaseSyntheticEvent) => Promise<void>;
+  isOpen: boolean;
 }
 
-// This component gets unmounted when config is null so no need to reset the state.
 function KnowledgeConfigurationSheetContent({
   action,
   config,
   onSave,
   control,
   handleSubmit,
+  isOpen,
 }: KnowledgeConfigurationSheetContentProps) {
   const { owner } = useAgentBuilderContext();
   const { supportedDataSourceViews } = useDataSourceViewsContext();
   const { spaces } = useSpacesContext();
+  const { mcpServerViewsWithKnowledge, isMCPServerViewsLoading } =
+    useMCPServerViewsContext();
 
   const instructions = useWatch<AgentBuilderFormData, "instructions">({
     name: "instructions",
   });
 
-  const [currentPageId, setCurrentPageId] = useState<ConfigurationSheetPageId>(
-    CONFIGURATION_SHEET_PAGE_IDS.DATA_SOURCE_SELECTION
+  const initialPageId =
+    action && isSupportedAgentBuilderAction(action)
+      ? CONFIGURATION_SHEET_PAGE_IDS.CONFIGURATION
+      : CONFIGURATION_SHEET_PAGE_IDS.DATA_SOURCE_SELECTION;
+
+  const [currentPageId, setCurrentPageId] =
+    useState<ConfigurationSheetPageId>(initialPageId);
+
+  useEffect(() => {
+    if (action) {
+      if (isSupportedAgentBuilderAction(action)) {
+        setCurrentPageId(CONFIGURATION_SHEET_PAGE_IDS.CONFIGURATION);
+        const serverName = ACTION_TYPE_TO_MCP_SERVER_MAP[action.type];
+        setSelectedMCPServerName(serverName);
+        if (serverName && isKnowledgeServerName(serverName)) {
+          setDynamicConfig(CAPABILITY_CONFIGS[serverName]);
+        }
+      }
+    }
+  }, [action]);
+
+  useEffect(() => {
+    if (!isOpen && !action) {
+      setCurrentPageId(CONFIGURATION_SHEET_PAGE_IDS.DATA_SOURCE_SELECTION);
+      setSelectedMCPServerName("search");
+      setDynamicConfig(CAPABILITY_CONFIGS["search"]);
+    }
+  }, [isOpen, action]);
+
+  const [selectedMCPServerName, setSelectedMCPServerName] = useState<
+    string | null
+  >(
+    action && isSupportedAgentBuilderAction(action)
+      ? ACTION_TYPE_TO_MCP_SERVER_MAP[action.type]
+      : "search"
+  );
+  const [dynamicConfig, setDynamicConfig] = useState<CapabilityConfig | null>(
```

Comment:
Why not having only one config state in `KnowledgeConfigurationSheet` and pass `setConfig` to `KnowledgeConfigurationSheetContent`?